libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt

```diff
@@ -0,0 +1,67 @@
+package com.mapbox.navigation.ui.voice.api
+
+import com.mapbox.api.directions.v5.models.VoiceInstructions
+import com.mapbox.navigation.utils.internal.logW
+
+internal class TimeBasedNextVoiceInstructionsProvider(
+    private val observableTimeSeconds: Int,
+) : NextVoiceInstructionsProvider {
+
+    override fun getNextVoiceInstructions(progress: RouteProgressData): List<VoiceInstructions> {
```

Comment:
The implementation details are in the doc references in the ticket.
TLDR: yes, it will be called on RouteProgress updates but not on every update. By default the interval will be `0.5 * observableTimeSeconds`.
The same instructions will not be downloaded twice but it's not for this class to handle.
I see your point about duplicated work here but I'm something can or should be done. For 2 reasons:
1) There's relatively little work here: it's not downloading, file reading or sth like that;
2) The instructions will be downloaded in parallel so we don't know if the latest downloaded instruction is the last one or not. Even if it is, what if some previous instructions failed to be downloaded? I think we'd want to retry the request then. So yeah, before making the request already downloaded stuff will definitely be filtered out but I don't think any changes are required here.